diff -Naur pidgin-2.3.0.orig/libpurple/conversation.c pidgin-2.3.0/libpurple/conversation.c
--- pidgin-2.3.0.orig/libpurple/conversation.c	2007-11-24 20:05:52.000000000 +0100
+++ pidgin-2.3.0/libpurple/conversation.c	2007-11-27 17:28:32.000000000 +0100
@@ -2201,6 +2201,13 @@
 										PURPLE_SUBTYPE_CONVERSATION),
 						 purple_value_new(PURPLE_TYPE_UINT));
 
+	purple_signal_register(handle, "blocked-im-msg",
+						 purple_marshal_VOID__POINTER_POINTER,
+						 NULL, 2,
+						 purple_value_new(PURPLE_TYPE_SUBTYPE,
+										PURPLE_SUBTYPE_ACCOUNT),
+						 purple_value_new_outgoing(PURPLE_TYPE_STRING));
+
 	purple_signal_register(handle, "writing-chat-msg",
 						 purple_marshal_BOOLEAN__POINTER_POINTER_POINTER_POINTER_UINT,
 						 purple_value_new(PURPLE_TYPE_BOOLEAN), 5,
diff -Naur pidgin-2.3.0.orig/libpurple/server.c pidgin-2.3.0/libpurple/server.c
--- pidgin-2.3.0.orig/libpurple/server.c	2007-11-24 20:06:01.000000000 +0100
+++ pidgin-2.3.0/libpurple/server.c	2007-11-27 17:28:32.000000000 +0100
@@ -554,8 +554,11 @@
 
 	if (PURPLE_PLUGIN_PROTOCOL_INFO(gc->prpl)->set_permit_deny == NULL) {
 		/* protocol does not support privacy, handle it ourselves */
-		if (!purple_privacy_check(account, who))
+		if (!purple_privacy_check(account, who)) {
+			purple_signal_emit(purple_conversations_get_handle(),
+					"blocked-im-msg", account, &who);
 			return;
+		}
 	}
 
 	/*
@@ -774,8 +777,11 @@
 	account = purple_connection_get_account(gc);
 	if (PURPLE_PLUGIN_PROTOCOL_INFO(gc->prpl)->set_permit_deny == NULL) {
 		/* protocol does not support privacy, handle it ourselves */
-		if (!purple_privacy_check(account, who))
+		if (!purple_privacy_check(account, who)) {
+			purple_signal_emit(purple_conversations_get_handle(),
+					"blocked-im-msg", account, &who);
 			return;
+		}
 	}
 
 	plugin_return = GPOINTER_TO_INT(purple_signal_emit_return_1(
